packages/@lwc/engine-dom/src/renderer.ts

```diff
@@ -126,11 +116,15 @@ export function nextSibling(node: Node): Node | null {
     return node.nextSibling;
 }
 
-export function attachShadow(element: Element, options: ShadowRootInit): ShadowRoot {
-    if (hydrating) {
-        return element.shadowRoot!;
+export function attachShadow(element: HTMLElement, options: ShadowRootInit): ShadowRoot {
+    let internals: ElementInternals | undefined;
+    try {
+        internals = element.attachInternals?.();
+    } catch {
+        // unregistered elements will throw, in which case we go with plan B
+        // to inspect the open shadowRoot on the element.
     }
-    return element.attachShadow(options);
+    return internals?.shadowRoot || element.shadowRoot || element.attachShadow(options);
```

Comment:
I would suggest replacing this function with:

```js
    if (!isNull(element.shadowRoot)) {
        return element.shadowRoot;
    }
    return element.attachShadow(options);
```

The `attachInternals()` is not strictly related to this PR and can be handled in a follow-up PR.

Reply:
I agree to do the internals on another PR.